File: sem 4/JIPP/lab2/zad1/ModuloField.cpp
#include "ModuloField.h"
#include "CharacteristicException.h"

ModuloField::ModuloField()
{
}

ModuloField::ModuloField(long value)
{
    if (value >= characteristic || value < 0)
    {
        throw CharacteristicException("Value must be in range [0, characteristic).");
    }
    this->value = value;
}

bool ModuloField::operator==(const ModuloField &other) const
{
    return value == other.value;
}

bool ModuloField::operator!=(const ModuloField &other) const
{
    return value != other.value;
}

bool ModuloField::operator<(const ModuloField &other) const
{
    return value < other.value;
}

bool ModuloField::operator<=(const ModuloField &other) const
{
    return value <= other.value;
}

bool ModuloField::operator>(const ModuloField &other) const
{
    return value > other.value;
}

bool ModuloField::operator>=(const ModuloField &other) const
{
    return value >= other.value;
}

ModuloField ModuloField::operator=(const ModuloField &other)
{
    this->value = other.value;
    return *this;
}

ModuloField ModuloField::operator+(const ModuloField &other) const
{
    return ModuloField(value + other.value);
}

ModuloField &ModuloField::operator+=(const ModuloField &other)
{
    this->value += other.value;
    this->value %= characteristic;
    return *this;
}

ModuloField ModuloField::operator-(const ModuloField &other) const
{
    return ModuloField((value - other.value + characteristic) % characteristic);
}

ModuloField &ModuloField::operator-=(const ModuloField &other)
{
    this->value -= other.value;
    this->value = (this->value + characteristic) % characteristic;
    return *this;
}

ModuloField ModuloField::operator*(const ModuloField &other) const
{
    return ModuloField(value * other.value % characteristic);
}

ModuloField &ModuloField::operator*=(const ModuloField &other)
{
    this->value *= other.value;
    this->value %= characteristic;
    return *this;
}

ModuloField ModuloField::operator/(const ModuloField &other) const
{
    return ModuloField(value * other.getInverse() % characteristic);
}

ModuloField &ModuloField::operator/=(const ModuloField &other)
{
    this->value *= other.getInverse();
    this->value %= characteristic;
    return *this;
}

ModuloField::operator long() {
    return value;
}

long ModuloField::getValue() const
{
    return value;
}

long ModuloField::getCharacteristic()
{
    return characteristic;
}

long ModuloField::getInverse() const
{
    long a = value;
    long b = characteristic;
    long x = 0;
    long y = 1;
    long u = 1;
    long v = 0;

    while (a != 0)
    {
        long r = b % a;
        long q = b / a;
        long m = x - u * q;
        long n = y - v * q;

        b = a;
        a = r;
        x = u;
        y = v;
        u = m;
        v = n;
    }

    return (x % characteristic + characteristic) % characteristic;
}

std::ostream &operator<<(std::ostream &os, const ModuloField &m)
{
    os << m.getValue();
    return os;
}

std::istream &operator>>(std::istream &is, ModuloField &m)
{
    is >> m.value;
    return is;
}

File: sem 4/JIPP/lab3/zad3/ModuloField.py
from CharacteristicException import CharacteristicException


class ModuloField:
    characteristic = 1234577

    def __init__(self, value):
        if value >= ModuloField.characteristic or value < 0:
            raise CharacteristicException("Value must be in range [0, characteristic).")
        self.value = value

    def __eq__(self, other):
        return self.value == other.value

    def __ne__(self, other):
        return self.value != other.value

    def __lt__(self, other):
        return self.value < other.value

    def __le__(self, other):
        return self.value <= other.value

    def __gt__(self, other):
        return self.value > other.value

    def __ge__(self, other):
        return self.value >= other.value

    def __add__(self, other):
        return ModuloField((self.value + other.value) % ModuloField.characteristic)

    def __iadd__(self, other):
        self.value = (self.value + other.value) % ModuloField.characteristic
        return self

    def __sub__(self, other):
        return ModuloField((self.value - other.value + ModuloField.characteristic) % ModuloField.characteristic)

    def __isub__(self, other):
        self.value = (self.value - other.value + ModuloField.characteristic) % ModuloField.characteristic
        return self

    def __mul__(self, other):
        return ModuloField((self.value * other.value) % ModuloField.characteristic)

    def __imul__(self, other):
        self.value = (self.value * other.value) % ModuloField.characteristic
        return self

    def __truediv__(self, other):
        return ModuloField((self.value * self.get_inverse(other.value)) % ModuloField.characteristic)

    def __itruediv__(self, other):
        self.value = (self.value * self.get_inverse(other.value)) % ModuloField.characteristic
        return self

    def get_value(self):
        return self.value

    @staticmethod
    def get_characteristic():
        return ModuloField.characteristic

    def get_inverse(self, value):
        a = value
        b = ModuloField.characteristic
        x, y, u, v = 0, 1, 1, 0

        while a != 0:
            r = b % a
            q = b // a
            m = x - u * q
            n = y - v * q

            b = a
            a = r
            x = u
            y = v
            u = m
            v = n

        return (x % ModuloField.characteristic + ModuloField.characteristic) % ModuloField.characteristic